aten/src/ATen/native/sparse/SparseTensor.cpp

```diff
@@ -744,6 +746,72 @@ SparseTensor _coalesce_sparse_cpu(const SparseTensor& self) {
 }
 
 DEFINE_DISPATCH(sparse_mask_intersection_out_stub);
+DEFINE_DISPATCH(sparse_mask_projection_out_stub);
+
+using OptTensor = c10::optional<Tensor>;
+
+std::tuple<Tensor, Tensor, OptTensor> sparse_mask_like_prepare_sparse_inputs(
+    const std::string& method_name,
+    const Tensor& t,
+    const Tensor& mask) {
+  // This is a helper function for operations that implement "sparse_mask"-like
+  // functionality, namely, projection of values of one tensor onto the other.
+  // These operations mostly rely on COO intersection primitives that heavily
+  // exploit coalesced inputs to avoid any syncs and calls to sort. The problem
+  // is that these primitives might project first argument onto second one or
+  // the other way around depending on which arguments are coalesced and which are
+  // larger. This function prepares inputs for `sparse_mask` such that `t` is
+  // projected onto `mask` by sorting `t` if uncoalesced and artifically marking it
+  // as coalesced all while `mask` is set to uncoalesced.
+  // The result of this projectionk is going to be uncoalesced, so it is up to the
```

Comment:
Thanks for adding this comment!

This function returns three tensors. How do these relate to the inputs and to the result of the projection?

Reply:
IIUC, for
```python
lhs, rhs, lhs_hash_opt = sparse_mask_like_prepare_sparse_inputs(t, mask)
```
we have the following invariants
```python
lhs.indices() == t.indices()[lhs_hash]
lhs.values() == t.values()[lhs_hash]
rhs == mask
```
where `lhs_hash = lhs_hash_opt or slice(None)`, `lhs` may be a copy or a view of `t`, `rhs` is uncoaleced view of `mask`.

Reply:
See the comment in https://github.com/pytorch/pytorch/pull/94991#discussion_r1177575459.